doc/Maintenance_and_Troubleshooting.md

```diff
@@ -113,3 +113,64 @@ or have a look at the nodes:
 If you fall into some Kubernetes specific issues after a successful cluster
 creation, go through the official [Kubernetes](https://kubernetes.io/docs/tasks/debug/debug-cluster/)
 troubleshooting guide.
+
+## Etcd maintenance
+
+[Etcd](https://etcd.io/) is a highly-available key value store used as Kubernetes'
+backing store for all cluster data. This section contains etcd related maintenance
+notes from SCS k8s-cluster-api-provider project perspective.
+
+For further information about etcd maintenance visit an official [etcd maintenance guide](https://etcd.io/docs/v3.5/op-guide/maintenance/)
+and/or [Kubernetes etcd operating guide](https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/).
+
+### Defragmentation and backup
+
+Etcd storage can become fragmented over time, which causes the performance to decrease
+and may even cause premature out-of-space conditions. For this we had included a
```

Comment:
> which causes the performance to decrease
> and may even cause premature out-of-space conditions

I think these claims should be backed by referring to etcd documentation or some case study.

[etcd defragmentation docs](https://etcd.io/docs/v3.5/op-guide/maintenance/#defragmentation) AFAIU only say that while defragmentation is useful for giving back disk space to the host filesystem, the space is still available to etcd - without notions about performance or out-of-space conditions.

Let's back this up with some references or remove it.

(Of course, doing defragmentation before doing a file-based backup would have advantages like smaller backup sizes)

Reply:
> If the defragmentation is not performed in a long run the out-of-space condition may occur on the host filesystem which may lead to performance degradation on both host and then potentially on etcd too. I did not find any shiny case study that shows the correlation between the host's out-of-space and the host's performance degradation as this could depend on many factors like the type of disk etc.

Well, if the disk really runs out of space, this may cause indeed `unpredictable etcd cluster behavior`, not just performance problems.

> It is true that defragmentation is useful for giving back disk space to the host filesystem. If the defragmentation is not performed in a long run the out-of-space condition may occur on the host filesystem

That implies that etcd does not reuse space for itself, causing a steady increase of disk space usage. That would be a contradiction to the content of the etcd documentation.

Under https://etcd.io/docs/v3.5/op-guide/maintenance/#space-quota, the docs say that defragmentation is required when the cluster ran out of space earlier - to bring the cluster in an operable state again.
But even if doing periodic defragmentation, there is AFAIU nothing preventing the keyspace to explode *between the defragmentation cycles*.

According to my understanding, reasonable sizings/rate-limits/compactions/quotas are required to prevent out-of-space conditions as well as possible; with defragmentation playing mostly a minor role when recovering/backing-up clusters.

Reply:
I got your point that defragmentation does not prevent out-of-space conditions, but mainly compactions should be performed before actual defragmentation. Also, it could be unclear that etcd reuses space for itself. Therefore, I removed the statement from docs 41d003bba43d0362a93e369f911f7421769b3a8c.

Reply:
Thanks!

> I got your point that defragmentation (alone) does not prevent out-of-space conditions, but compaction + defragmentation do this.

(Even though I'd still argue that AFAIU, defragmentation does not really prevent out-of-space conditions *at all*, unless other things on the same disk use space that should have been planned to be available to etcd)

Reply:
@joshmue - I agree that we should also start to look at rate limits ... if we want to substantially increase the operational safety.